test/cql-pytest/test_permissions.py

```diff
@@ -39,3 +40,89 @@ def test_grant_applicable_data_and_role_permissions(cql, test_keyspace, cassandr
             for permission in all_permissions.difference(set(permissions)):
                 with pytest.raises((InvalidRequest, SyntaxException), match="support.*permissions"):
                     cql.execute(f"GRANT {permission} ON {resource} TO {user}")
+
+
+def eventually_authorized(fun, timeout_s=10):
+    for i in range(timeout_s * 10):
+        try:
+            return fun()
+        except Unauthorized as e:
+            time.sleep(0.1)
+    return fun()
+
+def eventually_unauthorized(fun, timeout_s=10):
+    for i in range(timeout_s * 10):
+        try:
+            fun()
+            time.sleep(0.1)
+        except Unauthorized as e:
+            return
+    try:
+        fun()
+        pytest.fail(f"Function {fun} was not refused as unauthorized")
+    except Unauthorized as e:
+        return
+
+def grant(cql, permission, resource, username):
+    cql.execute(f"GRANT {permission} ON {resource} TO {username}")
+
+def revoke(cql, permission, resource, username):
+    cql.execute(f"REVOKE {permission} ON {resource} FROM {username}")
+
+# Helper function for checking that given `function` can only be executed
+# with given `permission` granted, and returns an unauthorized error
+# with the `permission` revoked.
+def check_enforced(cql, username, permission, resource, function):
+    grant(cql, permission, resource, username)
```

Comment:
I suggested you could also check that it doesn't work before the grant. If I understand correctly, the user should not have any rights before the grant, right?